src/WebJobs.Script/ScriptHostBuilderExtensions.cs

```
@@ -296,12 +300,49 @@ public static IHostBuilder AddScriptHostCore(this IHostBuilder builder, ScriptAp
                     services.TryAddEnumerable(ServiceDescriptor.Singleton<IHostedService, FunctionsScaleMonitorService>());
                 }
                 services.TryAddSingleton<FunctionsScaleManager>();
+
+                services.AddHostOverrides();
             });
 
             RegisterFileProvisioningService(builder);
             return builder;
         }
 
+        public static void AddHostOverrides(this IServiceCollection services)
+        {
+            // Custom implementations that the Host overrides
+            services.AddSingleton<ScheduleMonitor, AzureStorageScheduleMonitor>();
+            services.AddSingleton<IDistributedLockManager>(provider => AddLockManager(provider));
+        }
+
+        public static IDistributedLockManager AddLockManager(IServiceProvider provider)
+        {
+            try
+            {
+                var azureStorageProvider = provider.GetRequiredService<IAzureStorageProvider>();
+                var loggerFactory = provider.GetRequiredService<ILoggerFactory>();
+                var container = azureStorageProvider.GetBlobContainerClient();
+
+                if (container != null)
+                {
+                    return new BlobLeaseDistributedLockManager(loggerFactory, azureStorageProvider);
+                }
+            }
+            catch (Exception ex)
+            {
+                var str = ex.ToString();
+            }
+
+            return new InMemoryDistributedLockManager();
```

Is this fallback to the inmemory manager just for testing? Is it possible we'll hit this codepath in prod, and the fact that its silently falling back will be confusing? Should we emit a log entry about the fallback?

---

This is intended behavior for cases where Blob storage isn't available (i.e. a placeholder that is warmed up will use this InMemoryDistributedLockManager. Don't think we should log here because it is intended and may just cause more noise.

---

Okay do you know what exceptions to check for here? ArgumentException maybe? If you can scope down the catch block to just handling the expected exception type I'd feel more comfortable with this.

---

Still not clear to me that we need to pull this type into host. SDK will use it's version of InMemoryDistributedLockManager IFF we haven't provided our own implementation, so seems we can just rely on that?

---

In order to determine if Blob Storage is available, we need access to the IAzureStorageProvider service. When the scriptHost has a parent scope (i.e. WebHost), IAzureStorageProvider is not available yet when these services are registered, so I used the AddSingleton<TService>(this IServiceCollection services, Func<IServiceProvider, TService> implementationFactory) so that the factory can create the implementation at runtime when it is needed. This factory requires an implementation to be returned, so I brought in InMemoryDistributedLockManger. 

Do let me know if there's a way around this.